Homework BigInteger

asminimulin opened the pull request.

Commit:
lecture2: added

asminimulin committed:
Add .gitignore

asminimulin committed:
Add .clang-format

asminimulin committed:
Implement BigInteger

asminimulin committed:
Add warnings and set C++ standard version

asminimulin committed:
Fix missing namespace using delcaration

asminimulin committed:
Fix toString() method when value is equal to "0"

asminimulin committed:
Improve performance

asminimulin committed:
Remove redundant memory allocation in multiplication operation

asminimulin committed:
Fix merge conflict from upstream/master

asminimulin committed:
Merge branch 'master' into hw-BigInteger
Merge new tests into hw-branch

asminimulin committed:
Add unary plus

asminimulin committed:
Fix: remove redundant method SummarizeAbsoluteValuesInplace

asminimulin committed:
fix buggy math at BigInteger::ShiftBitsRight

asminimulin committed:
Add sanitizers to CMake build

lgAshanl commented on module-1/homework/BigInteger/biginteger.cpp:
```diff
@@ -0,0 +1,594 @@
+
+big_integer::BigInteger& big_integer::BigInteger::operator+=(
+    const big_integer::BigInteger& other) {
+  auto sum = *this + other;
```
Содаешь дополнительный BigInteger на стеке: лишний вызов конструктора и деструктора. Ну и памяти на стеке больше используешь. Напиши изменение левого операнда на месте

lgAshanl commented on module-1/homework/BigInteger/biginteger.cpp:
```diff
@@ -0,0 +1,594 @@
+
+bool big_integer::BigInteger::operator<(
+    const big_integer::BigInteger& other) const noexcept {
+  auto compare_result = CompareAbsoluteValues(this->limbs_, other.limbs_);
```
Кажется, тут и далее проблема со знаками

lgAshanl commented on module-1/homework/BigInteger/biginteger.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+big_integer::BigInteger& big_integer::BigInteger::operator++() {
+  *this = *this + BigInteger(1);
```
То же, что и с +=

lgAshanl commented on module-1/homework/BigInteger/biginteger.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+std::string big_integer::BigInteger::toString() const {
+  if (IsZeroed()) return "0";
+  std::string result;
+  if (is_negative_) {
+    result += '-';
+  }
+  std::vector<uint32_t> limbs = limbs_;
+  for (;;) {
+    if (limbs.size() == 1U && limbs[0] == 0) break;
+    auto division_result = DivideAbsoluteValueByLimb(limbs, 10U);
+    result += char(division_result.second + '0');
+    std::swap(limbs, division_result.first);
```
Вот в таких местах очень явно код делится на этапы: проверка на ноль -> первичная инициализация результата -> обработка в цикле(этот этап тоже в свою очередь делится на этапы). В таких случаях хорошо этапы дополнительно разделять пустой строкой: это кратно повышает читаемость (как с абзацами в текстах). Таких мест вообще много, тут оно просто явное. Проверь, пожалуйста, весь код.

lgAshanl commented on module-1/homework/BigInteger/biginteger.cpp:
```diff
@@ -0,0 +1,594 @@
+
+void big_integer::BigInteger::ShiftBitsLeft(std::vector<uint32_t>& limbs,
+                                            size_t count) {
+  if (count == 0) return;
```
А return стоит переносить на другую строку

lgAshanl commented on module-1/homework/BigInteger/biginteger.cpp:
```diff
@@ -0,0 +1,594 @@
+}
+
+bool big_integer::BigInteger::IsZeroed() const {
+  return limbs_.size() == 1 && limbs_[0] == 0;
```
Просто совет: стоит оборачивать в скобочки операнды, иначе, в случае ошибки по приоритетам операций, дебажить замучаешься

lgAshanl commented on module-1/homework/BigInteger/biginteger.h:
```diff
@@ -0,0 +1,173 @@
+    EQUAL = 0,
+    GREATER = 1,
+  };
+  static CompareResult CompareAbsoluteValues(
+      const std::vector<uint32_t>& left, const std::vector<uint32_t>& right);
+  static std::vector<uint32_t> SummarizeAbsoluteValues(
+      const std::vector<uint32_t>& left, const std::vector<uint32_t>& right);
+  static std::vector<uint32_t> SubtractAbsoluteValues(
+      const std::vector<uint32_t>& left, const std::vector<uint32_t>& right);
+  static std::vector<uint32_t> MultiplyAbsoluteValues(
+      const std::vector<uint32_t>& left, const std::vector<uint32_t>& right);
+  static std::vector<uint32_t> MultiplyAbsoluteValuesWithKaratsubaAlgo(
+      const std::vector<uint32_t>& left, const std::vector<uint32_t>& right);
+  static std::vector<uint32_t> MultiplyAbsoluteValuesNaive(
+      const std::vector<uint32_t>& shorter,
+      const std::vector<uint32_t>& longer);
+  static std::pair<std::vector<uint32_t>, std::vector<uint32_t>>
+  DivideAbsoluteValues(const std::vector<uint32_t>& left,
+                       const std::vector<uint32_t>& right);
+  static std::pair<std::vector<uint32_t>, uint32_t> DivideAbsoluteValueByLimb(
+      const std::vector<uint32_t>& left, uint32_t limb);
+  static void ShiftLimbsLeft(std::vector<uint32_t>& limbs, size_t count);
+  static void ShiftLimbsRight(std::vector<uint32_t>& limbs, long count);
+  static void ShiftBitsLeft(std::vector<uint32_t>& limbs, size_t count);
+  static void ShiftBitsRight(std::vector<uint32_t>& limbs, long count);
+  static constexpr uint32_t GetLeastBitsMask(size_t bits_count);
+  static void Normalize(std::vector<uint32_t>& limbs) noexcept;
+  static size_t GetMostSignificantBitPosition(uint32_t value) noexcept;
+  static size_t GetMostSignificantBitPosition(
+      const std::vector<uint32_t>& value) noexcept;
```
Тут тоже явно не хватает доп. строк между объявлениями

asminimulin committed:
fix codestyle: add spacing between function declaration in header

asminimulin committed:
fix buggy comparing operations: handle the sign of number properly

asminimulin committed:
fix: optimize arithmetic operations to avoid unnecessary Biginteger constructor and destructors calls

asminimulin committed:
refactor BigInteger: remove templates

Make operator bool() explicit
Also some readability improvements have done

asminimulin committed:
codestyle: add parentheses in if statements and add some blank lines

asminimulin committed:
codestyle: place return on the next line

lgAshanl commented on module-1/homework/BigInteger/biginteger.cpp:
```diff
@@ -0,0 +1,771 @@
+std::vector<uint32_t> big_integer::BigInteger::SummarizeAbsoluteValues(
+    const std::vector<uint32_t>& left, const std::vector<uint32_t>& right) {
+
+  std::vector<uint32_t> result(std::max(left.size(), right.size()));
+
```
Ты, в частности, используешь эту функцию для += и -=. И создаешь доп. вектор result. Хотя мог бы не выделять его, а модифицировать сразу left. Т.е. у тебя в += и -= всё равно создается доп копия, но уже не BigInt, а std::vector, но принципиально мало, что поменялось

asminimulin committed:
fix: update codestyle rules and add some 'fixme's

asminimulin committed:
feature: Use LimbType and ExtendedLimbType. Precalculate limb_type_size_.

Replace hardcoded uint32_t with LimbType and hardcoded uint64_t with ExtendedLimbType. Now limb type can be easily replaced with any integer type only in header file.
Also improve codestyle and add todo

asminimulin committed:
feature(biginteger): implement inplace sum and subtraction and use it where possible

asminimulin committed:
fix(biginteger): implement fixmes

asminimulin committed:
tests(biginteger): add tests for comparing negative values
